torch/csrc/jit/codegen/cuda/fusion_executor.cpp

```diff
@@ -0,0 +1,589 @@
+
+  auto fusion_inputs = fusion_.inputs();
+  EvaluationContext ec(&fusion_);
+
+  // This should probably move to EvaluationContext as we may want to bind input
+  // values frequently. Bind fusion input values to runtime values.
+  for (size_t i = 0; i < fusion_.inputs().size(); i++) {
+    if (fusion_.inputs()[i]->getValType() == ValType::TensorView) {
+      TensorView* cg_tensor = fusion_.inputs()[i]->as<TensorView>();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_inputs[i].isTensor(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      auto aten_tensor = aten_inputs[i].toTensor();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_tensor.ndimension() == cg_tensor->getRootDomain().size(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      for (size_t dim = 0; dim < cg_tensor->getRootDomain().size(); dim++) {
+        ec.bind(
+            cg_tensor->getRootDomain()[dim]->extent(),
+            aten_tensor.sizes()[dim]);
+      }
+    }
+  }
+
+  LaunchParams lp;
+
+  // Grab all used values and run through them
+  auto unordered_vals = DependencyCheck::getAllValsBetween(
+      {fusion_inputs.begin(), fusion_inputs.end()}, fusion_.outputs());
+
+  // Grab only the TensorViews
+  std::set<TensorView*> unordered_tvs;
+  for (auto val : unordered_vals) {
+    if (val->getValType().value() == ValType::TensorView) {
+      TensorView* tv = val->as<TensorView>();
+      unordered_tvs.emplace(tv);
+    }
+  }
+
+  // Infer bindings
+  for (auto tv : unordered_tvs) {
+    for (auto id : tv->domain()->domain()) {
+      if (id->isThread()) {
+        auto val = ExpressionEvaluator::evaluate(id->rawExtent(), &ec);
+        TORCH_INTERNAL_ASSERT(
+            val,
+            "Tried to evaluate, ",
+            id,
+            ", within the tensor ",
+            tv,
+            " to set launch bounds but could not.");
+        lp.bind(val.value(), id->parallel_method());
+      }
+    }
+  }
+
+  return lp;
+}
+
+// This function is here for testing purposes only
+std::vector<at::Tensor> FusionExecutor::runFusion(
+    const at::ArrayRef<IValue> inputs,
+    const std::vector<at::Tensor>& outputs) {
+  TORCH_INTERNAL_ASSERT(
+      fusion_id > 0, "Cannot run fusion, it was not compiled.");
+
+  FusionGuard fg(&fusion_);
+
+  validateKernelArgs(&fusion_, inputs, outputs, options_.device);
+
+  const auto prior_device = at::cuda::current_device();
+  at::cuda::set_device(options_.device);
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_INTERNAL_ASSERT(!outputs.empty(), "No outputs set for test kernel.");
+
+  KernelArgumentHolder kah;
+  kah.appendArgs(inputs);
+  kah.appendArgs(outputs);
+
+  LaunchParams lp = computeLaunchParams(inputs);
+
+  if (has_random) {
+    const auto rand_offset =
+        4 * (std::ceil(outputs[0].numel() / (4.0 * 128 * lp.gdimx())) + 1);
+    kah.appendPhilox(rand_offset);
+  }
+
+  // TODO SUPPORT GRID REDUCTIONS:
```

Comment:
I was discussing with @naoyam on this topic earlier.

I think it would be cleaner if we can have aggregated global buffer request and codegen split buffer appropriately. (same way that we would work with dynamic shared memory.) That we we can infer buffer size in the IR easier using ExpressionEvaluator.

Reply:
I don't think we need split buffers, but we can allocate individual buffers. Yes, I am planning to do this.